### PR TITLE
Message format document rewrite.

### PR DESCRIPTION
**Pull request by boguslaw-wojcik**

### CODE REVIEW

**Comment by boguslaw-wojcik** [comment]
@chribru @GenixPL @eternalamenity @BudzikM 
Probably easier to read here:

image

image

**Review by eternalamenity — APPROVED** [review]
lgtm

**Comment by BudzikM** [comment]
Can we make corid mandatory/required ? Would it make sense ?

**Comment by BudzikM** [comment]
Why do we use uid and its even mandatory ?

File: contributing.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Contributing
+
+## Local Preview
+
+This documentation is utilizing [Docsify](https://docsify.js.org/#/) for rendering purposes. To preview the documentation locally you need to have Docsify CLI installed which in turn requires NPM package manager.
```
Comment (BudzikM) [inline contributing.md:5]: utilizes

**Comment by boguslaw-wojcik** [comment]
@BudzikM 

1) Correlation ID cannot be required - it makes sense only for reply to a request. So uncorrelated event should not have corid nor a command will have a corid.

2) UID is mandatory to be able to reference a particular message, first in corid field in a correlated response, second, in internal processing of messages.

File: message-format.md [hunk 2]
```diff
@@ -1,47 +1,106 @@
...
-ver      | String              | Yes      | Version of the message format, default: `"1"`.
-
-\*Required for Prime Fimp messages.
+# Message Format
+
+Messages sent over MQTT adhering to the FIMP protocol are JSON messages which meet schema requirements outlined in this document.
+
+## Example message
+
+Following is the example of a FIMP message containing energy consumption report from an electricity meter. More examples can be found in particular [services](services/services.md)
```
Comment (BudzikM) [inline message-format.md:7]: The following is an example

File: message-format.md [hunk 3]
```diff
@@ -1,47 +1,106 @@
...
+  "ctime": "2022-12-02T10:08:27.5+01:00",
+  "src": "zwave-ad",
+  "ver": "1",
+  "topic": "pt:j1/mt:evt/rt:dev/rn:zw/ad:1/sv:meter_elec/ad:7_0"
+}
+```
+
+## Message Properties
+
+Following is the list of message properties in accordance to FIMP version 1.
```
Comment (BudzikM) [inline message-format.md:36]: The following is a list...

Comment (BudzikM) [inline message-format.md:36]: in accordance with....

File: message-format.md [hunk 4]
```diff
@@ -1,47 +1,106 @@
...
+| `type`    | String       | `"evt.binary.report"`                     | Yes      | The name of the interface represented by the message. See the [Interface Format](#interface-format) section for more details.                                                                                                        |
+| `val_t`   | String       | `"bool"`                                  | Yes      | Type of the value stored in `val` property. See the list of [Value Types](#value-types) section for more details.                                                                                                                    |
+| `val`     | Any          | `3.33`                                    | Yes      | Payload of the message of which type is defined in `val_t` property.                                                                                                                                                                 |
```
Comment (BudzikM) [inline message-format.md:43]: 1) Payload of the message with type is.. (remove redundant 'of')
2) Any is not a type just placeholder. Maybe use __

File: message-format.md [hunk 5]
```diff
@@ -1,47 +1,106 @@
...
+| `storage` | Object       | `{"sub_value": "kWh"}`                    | No       | Defines optional storage aggregation key for multi-value reports. Applies only to device reports. Parser must accept missing key or `null` value as an empty `object`.                                                               |
+| `props`   | String Map   | `{"unit":"W"}`                            | No       | An optional map of properties. Please note that all properties including numeric ones must be encoded as strings. Parser must accept missing key or `null` value as an empty `map`.                                                  |
+| `tags`    | String Array | `{"test"}`                                | No       | An optional list of tags. Parser must accept missing key or `null` value as an empty `array`.                                                                                                                                        |
```
Comment (BudzikM) [inline message-format.md:46]: shall accept the missing key

File: message-format.md [hunk 6]
```diff
@@ -1,47 +1,106 @@
...
+| `resp_to` | String       | `"pt:j1/mt:rsp/rt:app/rn:angry-dog/ad:1"` | No       | A topic where a response to the request will be expected. Does not apply to most of the broadcast type of events, especially asynchronous device events for which adapter will ignore this property. Parser must accept missing key. |
```
Comment (BudzikM) [inline message-format.md:47]: A topic when a response to the request is expected.

Comment (BudzikM) [reply message-format.md:47]: ..for which adapter ignores this property.

Comment (BudzikM) [reply message-format.md:47]: parser shall accept the missing key.

File: message-format.md [hunk 7]
```diff
@@ -1,47 +1,106 @@
...
+| `uid`     | String       | `"9bb1be75-35d7-4069-ac00-b974315f7ec3"`  | Yes      | A unique message identifier in form of UUID. It is recommended to use Version 4.                                                                                                                                                     |
+| `ctime`   | String       | `"2022-12-02T10:08:27.5+01:00"`           | Yes      | Message creation time in RFC3339 format. For compatibility reasons a message parser must treat a missing key as the current timestamp. See [time format](#time-format) section for disambiguation.                                   |
+| `src`     | String       | `"smarthome-app"`                         | Yes      | Name of the message publisher identifying a HUB, mobile or cloud application or service. For compatibility reasons a message parser must tolerate a missing key.                                                                     |
```
Comment (BudzikM) [inline message-format.md:51]: mobile, cloud, application or a service.

File: message-format.md [hunk 8]
```diff
@@ -1,47 +1,106 @@
...
+| str_map     | `{"ip": "192.168.1.1"}`                                  |                                                                                                                             |
+| float_map   | `{"pi": 3.14}`                                           |                                                                                                                             |
+| bool_map    | `{"enabled": true}`                                      |                                                                                                                             |
+| object      | `{"nested": {"objects": "supported"}}`                   | Used for complex objects which cannot be mapped to primitive types. Structure is defined uniquely per service or interface. |
+| bin         | `"U28gbG9uZywgYW5kIHRoYW5rcyBmb3IgYWxsIHRoZSBmaXNoLg=="` | Used for `base64` encoded binary payloads.                                                                                  |
+
+## Time Format
+
+Time must be formatted in accordance to the RFC3339 format. A message parser must be able to read all the following timestamps as valid.
```
Comment (BudzikM) [inline message-format.md:78]: The timestamp shall be...

Comment (BudzikM) [reply message-format.md:78]: A message parser shall be

File: message-format.md [hunk 9]
```diff
@@ -1,47 +1,106 @@
...
+| Format                                   | Recommended |
+|------------------------------------------|-------------|
+| `"2006-01-02T15:04:05.999999999Z07:00"`  | Yes         | 
+| `"2006-01-02T15:04:05.999999999Z0700"`   | No          | 
+| `"2006-01-02 15:04:05.999999999 Z0700"`  | No          | 
+| `"2006-01-02 15:04:05.999999999 Z07:00"` | No          | 
+
+## Interface Format
+
+Each interface consist of three segments separated by a dot in accordance to this schema `{type}.{attribute}.{action}`:
```
Comment (BudzikM) [inline message-format.md:90]: consists of..
.. in accordance with..

File: message-format.md [hunk 10]
```diff
@@ -1,47 +1,106 @@
...
+
+| Segment     | Example      | Description                                                                                                                                                                         |
+|-------------|--------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| *type*      | `cmd`        | Defines the **type** of the interface, either `cmd` for consumed incoming commands or `evt` for produced outgoing events.                                                           |
+| *attribute* | `lvl`        | Defines the **attribute** or a **domain** of the service. A service might have multiple attributes, while each attribute might have multiple interfaces. Must utilize `snake_case`. |
```
Comment (BudzikM) [inline message-format.md:95]: Has to utilize

File: message-format.md (hunk 5 above)
Comment (boguslaw-wojcik) [reply message-format.md:46]: https://www.ietf.org/rfc/rfc2119.txt

File: message-format.md (hunk 6 above)
Comment (boguslaw-wojcik) [reply message-format.md:47]: Topic is a place where we expect the response, not time when we expect the response.

Must and shall are interchangeable in documentation in accordance with https://www.ietf.org/rfc/rfc2119.txt

The natural language differences between have to / must / shall do not apply to technical documentation, AFAIK.

File: message-format.md (hunk 7 above)
Comment (boguslaw-wojcik) [reply message-format.md:51]: Not really, but reworded to make it more intelligible.

File: message-format.md (hunk 10 above)
Comment (boguslaw-wojcik) [reply message-format.md:95]: Again https://www.ietf.org/rfc/rfc2119.txt

File: message-format.md [hunk 11]
```diff
@@ -1,47 +1,106 @@
...
+| `src`     | String       | `"smarthome-app"`                         | Yes      | Name of the message publisher identifying a HUB or mobile application or a cloud service. For compatibility reasons a message parser must tolerate a missing key.                                                                    |
+| `ver`     | String       | `"1"`                                     | Yes      | Version of the message format. The current version is `1`.                                                                                                                                                                           |
+| `topic`   | String       | `"pt:j1/mt:evt/rt:app/rn:vinculum/ad:1"`  | No       | The MQTT topic on which the message was originally received, required only when message is processed outside of MQTT context, for example in a Kafka stream.                                                                         |
```
Comment (GenixPL) [inline message-format.md:53]: I had no idea it was a part O.o

File: message-format.md [hunk 12]
```diff
@@ -1,47 +1,106 @@
...
+| `val_t`   | String       | `"bool"`                                  | Yes      | Type of the value stored in `val` property. See the list of [Value Types](#value-types) section for more details.                                                                                                                    |
+| `val`     | Any          | `3.33`                                    | Yes      | Payload of the message which type is defined in `val_t` property.                                                                                                                                                                    |
+| `storage` | Object       | `{"sub_value": "kWh"}`                    | No       | Defines optional storage aggregation key for multi-value reports. Applies only to device reports. Parser must accept missing key or `null` value as an empty `object`.                                                               |
```
Comment (GenixPL) [inline message-format.md:44]: what and why is this? I can see it's been added and we dont have in the app

File: message-format.md [hunk 13]
```diff
@@ -1,47 +1,106 @@
...
+| `props`   | String Map   | `{"unit":"W"}`                            | No       | An optional map of properties. Please note that all properties including numeric ones must be encoded as strings. Parser must accept missing key or `null` value as an empty `map`.                                                  |
+| `tags`    | String Array | `{"test"}`                                | No       | An optional list of tags. Parser must accept missing key or `null` value as an empty `array`.                                                                                                                                        |
```
Comment (GenixPL) [inline message-format.md:46]: list not object (`{"test"}`)

**Review by GenixPL — CHANGES_REQUESTED** [review]

File: message-format.md (hunk 12 above)
Comment (boguslaw-wojcik) [reply message-format.md:44]: It's for Vinculum only. You don't need to care about it.

File: message-format.md (hunk 11 above)
Comment (boguslaw-wojcik) [reply message-format.md:53]: It's for internal processing only.

File: message-format.md (hunk 13 above)
Comment (boguslaw-wojcik) [reply message-format.md:46]: Nice catch

**Review by GenixPL — APPROVED** [review]

**Pull request merged**